app/src/migration/java/org/mozilla/fenix/MigrationPushSubscriber.kt

```diff
@@ -0,0 +1,42 @@
+package org.mozilla.fenix
+
+import android.content.Context
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.flow.collect
+import mozilla.components.concept.push.PushService
+import mozilla.components.lib.state.ext.flowScoped
+import mozilla.components.support.migration.state.MigrationProgress
+import mozilla.components.support.migration.state.MigrationStore
+
+/**
+ * Migration-aware subscriber that disables the push service during an active migration
+ * and re-enables when complete.
+ */
+class MigrationPushSubscriber(
+    private val context: Context,
+    private val service: PushService,
+    private val store: MigrationStore
+) {
+    private var scope: CoroutineScope? = null
+
+    @UseExperimental(ExperimentalCoroutinesApi::class)
+    fun start() {
+        // Stop the service if it is already started.
+        service.stop()
+
+        // Observe for migration completed.
+        store.flowScoped { flow ->
+            flow.collect { state ->
+                if (state.progress == MigrationProgress.COMPLETED) {
+                    service.start(context)
+                }
+            }
+        }
+    }
+
+    fun stop() {
+        scope?.cancel()
```

Comment:
you never call this?

Reply:
Yeah, I wasn't sure where this call needed to be made, part of my questions, but I put up the PR since we can do this after.

Since we're running the application context, this would be unlikely to fail. It's still good to call this somewhere.

Reply:
Can you call this when you see `completed` event?

Reply:
We would need to cancel the scope from within the flow that we get the scope from which seems unsafe.

Spoke to Christian offline and he said we haven't done this any where before and can lead to a race. We do have cases where we don't ever try to cancel the scope, which might be whats best.